File: app/src/main/java/com/maciega/bartosz/crawl/DataActivity.java
package com.maciega.bartosz.crawl;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.SearchView;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.inputmethod.InputMethodManager;
import android.widget.ListView;

import com.maciega.bartosz.crawl.adapter.DataAdapter;
import com.maciega.bartosz.crawl.model.Url;
import com.maciega.bartosz.crawl.presenter.DataPresenter;

import java.util.List;

/**
 * Created by bartoszmaciega on 17/10/16.
 */

public class DataActivity extends AppCompatActivity implements DataPresenter.DataView, SearchView.OnQueryTextListener {

    ListView listView;
    DataAdapter dataAdapter;
    private DataPresenter presenter;
    private SearchView searchView;


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.data_activity);
        listView = (ListView) findViewById(R.id.list);
        presenter = new DataPresenter(this, this);
        dataAdapter = new DataAdapter(this);
        listView.setAdapter(dataAdapter);
        initToolbar();
    }


    private void initToolbar() {
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);
    }

    @Override
    protected void onResume() {
        super.onResume();
        load();
    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.search_menu, menu);
        MenuItem searchItem = menu.findItem(R.id.search);
        SearchView searchView = (SearchView) MenuItemCompat.getActionView(searchItem);
        searchView.setOnQueryTextListener(this);
        this.searchView = searchView;
        MenuItemCompat.setOnActionExpandListener(searchItem, searchExpandListener);

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            super.onBackPressed();
        }
        return super.onOptionsItemSelected(item);
    }

    private void load() {
        presenter.loadData();
    }


    private MenuItemCompat.OnActionExpandListener searchExpandListener = new MenuItemCompat.OnActionExpandListener() {
        @Override
        public boolean onMenuItemActionExpand(MenuItem item) {
            return true;
        }

        @Override
        public boolean onMenuItemActionCollapse(MenuItem item) {
            presenter.finishQuery();
            return true;
        }
    };


    @Override
    public void showUrls(List<Url> urls) {
        dataAdapter.setData(urls);
    }

    @Override
    public boolean onQueryTextSubmit(String query) {
        presenter.submitQuery(query);
        InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        imm.hideSoftInputFromWindow(searchView.getWindowToken(), 0);
        return true;
    }

    @Override
    public boolean onQueryTextChange(String newText) {

        return false;
    }
}

File: app/src/main/java/com/maciega/bartosz/crawl/DataActivity.kt
package com.maciega.bartosz.crawl

import android.content.Context
import android.os.Bundle
import android.support.v4.view.MenuItemCompat
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.SearchView
import android.support.v7.widget.Toolbar
import android.view.Menu
import android.view.MenuItem
import android.view.inputmethod.InputMethodManager
import android.widget.ListView

import com.maciega.bartosz.crawl.adapter.DataAdapter
import com.maciega.bartosz.crawl.model.Url
import com.maciega.bartosz.crawl.presenter.DataPresenter

/**
 * Created by bartoszmaciega on 17/10/16.
 */

class DataActivity : AppCompatActivity(), DataPresenter.DataView, SearchView.OnQueryTextListener {

    internal var listView: ListView? = null
    internal var dataAdapter: DataAdapter? = null
    private var presenter: DataPresenter? = null
    private var searchView: SearchView? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.data_activity)
        presenter = DataPresenter(this, this)
        dataAdapter = DataAdapter(this)
        listView = findViewById(R.id.list) as ListView
        listView!!.adapter = dataAdapter
        initToolbar()
    }


    private fun initToolbar() {
        val toolbar = findViewById(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.setDisplayShowHomeEnabled(true)
    }

    override fun onResume() {
        super.onResume()
        load()
    }


    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.search_menu, menu)
        val searchItem = menu.findItem(R.id.search)
        val searchView = MenuItemCompat.getActionView(searchItem) as SearchView
        searchView.setOnQueryTextListener(this)
        this.searchView = searchView
        MenuItemCompat.setOnActionExpandListener(searchItem, searchExpandListener)

        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            super.onBackPressed()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun load() {
        presenter!!.loadData()
    }


    private val searchExpandListener = object : MenuItemCompat.OnActionExpandListener {
        override fun onMenuItemActionExpand(item: MenuItem): Boolean {
            return true
        }

        override fun onMenuItemActionCollapse(item: MenuItem): Boolean {
            presenter!!.finishQuery()
            dataAdapter?.resetQuery()
            return true
        }
    }

    override fun onQueryTextSubmit(query: String): Boolean {
        presenter!!.submitQuery(query)
        val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(searchView!!.windowToken, 0)
        dataAdapter?.setQuery(query)
        return true
    }

    override fun onQueryTextChange(newText: String): Boolean {

        return false
    }

    override fun showUrls(urls: List<Url>) {
        dataAdapter?.setData(urls)
    }
}
